src/resmom/renew_creds.c

```diff
@@ -1093,153 +1185,357 @@ send_cred_sisters(job *pjob)
+}
 
-	numsup = getgroups(maxgroups, grplist);
+/**
+ * @brief
+ * 	singleshot_afslog - do imminent afslog, no process started
+ *
+ * @param[in] ticket - kerberos ticket
+ *
+ */
+void
+singleshot_afslog(struct krb_holder *ticket) {
+	if(k_hasafs()) {
+		k_setpag();
+		do_afslog(ticket->context, ticket->job_info);
+	}
+}
 
-	for (i = 0; i < numsup; ++i) {
-		/* last (4th) byte in pag is char 'A' */
-		if ((grplist[i] >> 24) == 'A') {
-			pag = grplist[i];
-			break;
+/**
+ * @brief
+ * 	do_afslog_on_signal - AFS log signal handler
+ *
+ * @param[in] signal - received signal
+ *
+ */
+static void
+do_afslog_on_signal(int signal) {
+	krb5_error_code ret;
+
+	if (signal == SIGHUP) {
+		if ((ret = do_afslog(afslog_ticket->context, afslog_ticket->job_info))) {
+			return;
 		}
+	} else {
+		rec_signal = signal;
 	}
+}
 
-	if (grplist)
-		free(grplist);
-
-	return pag;
+/**
+ * @brief
+ * 	wait_afslog - in AFS log process wait for signal HUP for do AFS log or
+ *	0 for terminate
+ *
+ * @param[in] signal - received signal
+ *
+ */
+static void
+wait_afslog() {
+	/* initialize signal catcher */
+	struct sigaction sa;
+	memset(&sa, 0, sizeof(sa));
+	sa.sa_handler = do_afslog_on_signal;
+	sigaction(SIGTERM, &sa, NULL);
+	sa.sa_handler = do_afslog_on_signal;
+	sigaction(SIGHUP, &sa, NULL);
+
+	while (rec_signal == 0) {
+		sleep(AFSLOG_TIME_SLEEP);
+	}
 }
 
 /**
  * @brief
- * 	setpag - if afs pag != 0 is provided then the pag is added to groups
- *	if the pag is not provided, a new pag is set
+ * 	start_afslog - fork AFS log process for a specific job task,
+ *	save PID file and do AFS log for the task. The AFS log need to be done
+ *	in the child of original process who set the pag.
  *
- * @param[in] pag - the pag group id
+ * @param[in] ptask - job task
+ * @param[in] ticket - kerberos ticket associated with process
+ * @param[in] fd1 - read socket - need to be closed in forked process
+ * @param[in] fd2 - write socker - need to be closed in forked process
  *
+ * @return 	int
+ * @retval	PBSGSS_OK on success
+ * @retval	!= PBSGSS_OK otherwise
  */
-void
-setpag(int32_t pag)
-{
-	gid_t *grplist = NULL;
-	int    numsup;
-	static int   maxgroups = 0;
+int
+start_afslog(const task *ptask, struct krb_holder *ticket, int fd1, int fd2) {
+	char buf[LOG_BUF_SIZE * 2] ;
+	char errbuf[LOG_BUF_SIZE];
+	int ret = PBS_KRB5_OK;
+	char pid_file[MAXPATHLEN];
+	int local_ticket = 0;
+
+	if(!k_hasafs())
+		return PBS_KRB5_OK;
 
-	if (k_hasafs() == 0)
-		return;
+	if (ptask == NULL)
+		return PBS_KRB5_ERR_INTERNAL;
 
-	if (pag == 0) {
-		k_setpag();
-		return;
+	job *pjob = ptask->ti_job;
+
+	if (*pjob->ji_qs.ji_fileprefix != '\0')
+		snprintf(pid_file, sizeof(pid_file), "%s%s_afslog_%8.8X.pid", path_jobs, pjob->ji_qs.ji_fileprefix, (unsigned int)ptask->ti_qs.ti_task);
+	else
+		snprintf(pid_file, sizeof(pid_file), "%s%s_afslog_%8.8X.pid", path_jobs, pjob->ji_qs.ji_jobid, (unsigned int)ptask->ti_qs.ti_task);
+
+	int fd = open(pid_file, O_CREAT|O_EXCL|O_WRONLY, 0600);
```

Comment:
Please declare new variables at the beginning of the code block, be it a function or otherwise.